### ISSUE TITLE
Undefined references when using TensorFlow Lite in Arduino Nano 33 BLE Sense + Platform IO

### ISSUE THREAD

**Issue by victorromeo** [tensorflow/tensorflow#40497]
@tensorflow/micro

**System information**
Ubunutu 20.04 on x86_64
TensorFlow from source 2.2.0
Target: Arduino Nano 33 BLE Sense

Hi,

I'm trying to get micro_speech running on a Arduino Nano 33 BLE Sense, using a model I've trained myself on Google Colab.  As Google Colab is currently using TensorFlow 2.2.0, this model is also 2.2.0.  Using the Arduino IDE 1.8.12 sample 'micro_speech', does work, and I can get inferences to occur, however this is a long way from TensorFlow 2.2.0.  In fact its many versions behind and as a result, I'm unable to get my model to run on this version.

In addition, this project is a stepping stone to another project which will swap out the model for other models, and introduce other libraries which introduce further development and debugging complexity.  For this reason, I've chosen to adopt the Platform IO + VSCode development environment.  With this combination I can debug the code, step by step, which is a feature not supported by the Microsoft VSCode Arduino extension or Arduino IDE itself.  I can also use unit testing technologies which aren't available elsewhere.

Unfortunately, my code is running into Undefined reference exceptions on build, in what is a fairly trivial port of the micro_speech application to the Platform IO code.  I get these same errors building the code using the local TensorFlow Arduino library zip which I built myself from the TensorFlow 2.2.0 master branch a couple of days ago.

May I please get some advice?
- Any ideas why Undefined references are appearing in local builds of [micro_speech](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/micro/examples/micro_speech), or my port of the [micro_speech](https://github.com/victorromeo/pio_micro_speech) product?  It was also reported in this issue [27629](https://github.com/tensorflow/tensorflow/issues/27629)
- Given the explosion of options, which development environment is the recommended approach for products which extend beyond 'hello-world' targeting micro-controllers? 
  - Arduino Online 
  - Arduino IDE with local TensorFlow Arduino library zip  
  - TensorFlow Lite + Bazel from command  line
  - Platform IO (IDE / VSCode / Atom)
  - Microsoft Arduino over VSCode
  - entirely manual arm-none-eabi-g++

```
> Executing task: platformio run --verbose <

Processing nano33ble (platform: nordicnrf52; board: nano33ble; framework: arduino; debug_tool: jlink; upload_protocol: jlink)
----------------------------------------------------------------
CONFIGURATION: https://docs.platformio.org/page/boards/nordicnrf52/nano33ble.html
PLATFORM: Nordic nRF52 4.2.1 > Arduino Nano 33 BLE
HARDWARE: NRF52840 64MHz, 256KB RAM, 960KB Flash
DEBUG: Current (jlink) External (cmsis-dap, jlink)
PACKAGES: 
 - framework-arduino-nrf52-mbedos 1.1.3 
 - tool-sreccat 1.164.0 (1.64) 
 - toolchain-gccarmnoneeabi 1.80201.181220 (8.2.1)
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Framework incompatible library /home/ian/.platformio/packages/framework-arduino-nrf52-mbedos/libraries/mbed-memory-status
Found 7 compatible libraries
More details about "Library Compatibility Mode": https://docs.platformio.org/page/librarymanager/ldf.html#ldf-compat-mode
Scanning dependencies...
Dependency Graph
|-- <PDM> 1.0 (/home/ian/.platformio/packages/framework-arduino-nrf52-mbedos/libraries/PDM)
|-- <micro_features> (/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/micro_features)
|   |-- <tensorflow> (/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow)
|-- <tensorflow> (/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow)
Building in release mode
arm-none-eabi-g++ -o .pio/build/nano33ble/firmware.elf -T linker_script.ld -DMBED_APP_SIZE=0xf0000 -DMBED_APP_START=0x10000 -DMBED_BOOT_STACK_SIZE=2048 -Wl,--gc-sections -Wl,--wrap,_calloc_r -Wl,--wrap,_free_r -Wl,--wrap,_malloc_r -Wl,--wrap,_memalign_r -Wl,--wrap,_realloc_r -Wl,--wrap,atexit -Wl,--wrap,exit -Wl,--wrap,main -Wl,-n -mcpu=cortex-m4 -mfloat-abi=softfp -mfpu=fpv4-sp-d16 -mthumb --specs=nano.specs --specs=nosys.specs -Wl,--as-needed .pio/build/nano33ble/src/audio_provider.cpp.o .pio/build/nano33ble/src/command_responder.cpp.o .pio/build/nano33ble/src/feature_provider.cpp.o .pio/build/nano33ble/src/main.cpp.o .pio/build/nano33ble/src/recognize_commands.cpp.o -L.pio/build/nano33ble -L/home/ian/.platformio/packages/framework-arduino-nrf52-mbedos/variants/ARDUINO_NANO33BLE -L/home/ian/.platformio/packages/framework-arduino-nrf52-mbedos/variants/ARDUINO_NANO33BLE/libs -Wl,--start-group -Wl,--whole-archive .pio/build/nano33ble/libe06/libPDM.a .pio/build/nano33ble/lib5e8/libtensorflow.a .pio/build/nano33ble/lib082/libmicro_features.a .pio/build/nano33ble/libFrameworkArduinoVariant.a .pio/build/nano33ble/libFrameworkArduino.a -lmbed -lcc_310_core -lcc_310_ext -lcc_310_trng -Wl,--no-whole-archive -lstdc++ -lsupc++ -lm -lc -lgcc -lnosys -Wl,--end-group
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/command_responder.cpp.o: in function `RespondToCommand(tflite::ErrorReporter*, long, char const*, unsigned char, bool)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/command_responder.cpp:48: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/feature_provider.cpp.o: in function `FeatureProvider::PopulateFeatureData(tflite::ErrorReporter*, long, long, int*)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/feature_provider.cpp:102: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/main.cpp.o: in function `tflite::MicroMutableOpResolver<4u>::GetOpDataParser(tflite::BuiltinOperator) const':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow/tensorflow/lite/micro/micro_mutable_op_resolver.h:62: undefined reference to `tflite::ParseOpData(tflite::Operator const*, tflite::BuiltinOperator, tflite::ErrorReporter*, tflite::BuiltinDataAllocator*, void**)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/main.cpp.o: in function `tflite::MicroMutableOpResolver<4u>::AddCustom(char const*, TfLiteRegistration*)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow/tensorflow/lite/micro/micro_mutable_op_resolver.h:98: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow/tensorflow/lite/micro/micro_mutable_op_resolver.h:108: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/main.cpp.o: in function `tflite::MicroMutableOpResolver<4u>::AddBuiltin(tflite::BuiltinOperator, TfLiteRegistration*)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow/tensorflow/lite/micro/micro_mutable_op_resolver.h:68: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/tensorflow/tensorflow/lite/micro/micro_mutable_op_resolver.h:78: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/main.cpp.o: in function `setup':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:75: undefined reference to `tflite::ops::micro::Register_DEPTHWISE_CONV_2D()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:80: undefined reference to `tflite::ops::micro::Register_FULLY_CONNECTED()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:85: undefined reference to `tflite::ops::micro::Register_SOFTMAX()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:90: undefined reference to `tflite::ops::micro::Register_RESHAPE()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:98: undefined reference to `tflite::MicroInterpreter::MicroInterpreter(tflite::Model const*, tflite::MicroOpResolver const&, unsigned char*, unsigned int, tflite::ErrorReporter*)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:102: undefined reference to `tflite::MicroInterpreter::AllocateTensors()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:109: undefined reference to `tflite::MicroInterpreter::input(unsigned int)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:131: undefined reference to `tflite::MicroInterpreter::~MicroInterpreter()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:59: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:114: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/main.cpp.o: in function `loop':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:172: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:157: undefined reference to `tflite::MicroInterpreter::Invoke()'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/main.cpp:164: undefined reference to `tflite::MicroInterpreter::output(unsigned int)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/main.cpp.o:(.data._ZZ5setupE20micro_error_reporter+0x0): undefined reference to `vtable for tflite::MicroErrorReporter'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/recognize_commands.cpp.o: in function `RecognizeCommands::ProcessLatestResults(TfLiteTensor const*, long, char const**, unsigned char*, bool*)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/recognize_commands.cpp:46: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/recognize_commands.cpp:60: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/src/recognize_commands.cpp:74: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/recognize_commands.cpp.o: in function `PreviousResultsQueue::push_back(PreviousResultsQueue::Result const&)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/include/recognize_commands.h:66: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/src/recognize_commands.cpp.o: in function `PreviousResultsQueue::pop_front()':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/include/recognize_commands.h:79: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/lib082/libmicro_features.a(micro_features_generator.cpp.o): in function `InitializeMicroFeatures(tflite::ErrorReporter*)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/micro_features/micro_features_generator.cpp:50: undefined reference to `FrontendPopulateState'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: /home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/micro_features/micro_features_generator.cpp:52: undefined reference to `tflite::ErrorReporter::Report(char const*, ...)'
/home/ian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/nano33ble/lib082/libmicro_features.a(micro_features_generator.cpp.o): in function `GenerateMicroFeatures(tflite::ErrorReporter*, short const*, int, int, signed char*, unsigned int*)':
/home/ian/Documents/PlatformIO/Projects/Nano33Take01/lib/micro_features/micro_features_generator.cpp:79: undefined reference to `FrontendProcessSamples'
collect2: error: ld returned 1 exit status
*** [.pio/build/nano33ble/firmware.elf] Error 1
================== [FAILED] Took 1.25 seconds ==================
The terminal process terminated with exit code: 1

Terminal will be reused by tasks, press any key to close it.
```

**Comment by victorromeo** [comment]
So, I've reverted the codebase back to the 'micro_speech' example code from the Arduino Web Editor, then transferred these into PlatformIO using VSCode.  The build is executing to completion now. Even if inferences are failing its progress.

I've abandoned the 'micro_speech' implementation as it current appears in the TensorFlow GitHub repo (v 2.2.0), as a source for this project, as it simply doesn't seem to be easily ported.

If I can recommend anything from my experience its that the example projects and the instructions for building them need to demonstrate the current implementation of the TensorFlow technologies, because that's where the online documentation for all other training is guiding developers.

The Arduino implementation I'm using is now targeting the TensorFlow Lite 2.1.x, trained using Keras over TensorFlow 2.3.0+ (tf-nightly), whereas micro_speech and speech_command examples are straight TF 1.x.

Happy for the issue to close at the discretion of the TF team.
